File: 04_02_2024_Minimum_Window_Substring.py
from collections import Counter

class Solution:
    def minWindow(self, s, t):
        if t == "":
            return ""
        mpt = Counter(t)
        len_s = len(s)
        pr = (-1, -1)
        mps = Counter()
        have = 0
        need = len(mpt)
        j = 0
        for i in range(len(s)):
            mps[s[i]] += 1
            if s[i] in mpt and mps[s[i]] == mpt[s[i]]:
                have += 1
            while have == need:
                if i - j + 1 <= len_s:
                    pr = (j, i)
                    len_s = i - j + 1
                mps[s[j]] -= 1
                if s[j] in mpt and mps[s[j]] < mpt[s[j]]:
                    have -= 1
                j += 1
        if pr[1] == -1 or pr[0] == -1:
            return ""
        return s[pr[0]:pr[1]+1]

File: 04_02_2024_Minimum_Window_Substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
       if (t == "") return "";
        unordered_map<char, int> mpt;
        for (auto& x : t) mpt[x]++;
        int len = s.size();
        pair<int, int> pr = {-1, -1};
        unordered_map<char, int> mps;
        int have = 0, need = mpt.size(), j = 0;
        for (int i = 0; i < s.size(); i++) {
            mps[s[i]]++;
            if (mpt.count(s[i]) && mps[s[i]] == mpt[s[i]]) {
                have++;
            }
            while (have == need) {
                if (i - j + 1 <= len) {
                    pr.first = j, pr.second = i;
                    len = i - j + 1;
                }
                mps[s[j]]--;
                if (mpt.count(s[j]) && mps[s[j]] < mpt[s[j]]) {
                    have--;
                }
                j++;
            }
        }
        if (pr.second == -1 || pr.first == -1) return "";
        string ans = "";
        for (int i = pr.first; i <= pr.second; i++) {
            ans.push_back(s[i]);
        }
        return ans;
    }
};
